src/analyses/cfg_dominators.h

```diff
@@ -44,124 +242,262 @@ class cfg_dominators_templatet
+      : cfg_dominators(cfg_dominators),
+        dfs_counter(0),
+        parent(cfg_dominators.cfg.size() + 1),
+        ancestor(cfg_dominators.cfg.size() + 1),
+        child(cfg_dominators.cfg.size() + 1),
+        vertex(cfg_dominators.cfg.size() + 1),
+        dom(cfg_dominators.cfg.size() + 1),
+        label(cfg_dominators.cfg.size() + 1),
+        semi(cfg_dominators.cfg.size() + 1),
+        size(cfg_dominators.cfg.size() + 1),
+        bucket(cfg_dominators.cfg.size() + 1)
+    {
+    }
 
-/// Computes the MOP for the dominator analysis
-template <class P, class T, bool post_dom>
-void cfg_dominators_templatet<P, T, post_dom>::fixedpoint(P &program)
-{
-  std::list<T> worklist;
+    void fixedpoint(P &program);
 
-  if(cfg.nodes_empty(program))
-    return;
+  private:
+    cfg_dominators_templatet &cfg_dominators;
+    node_indext dfs_counter;
+    std::vector<node_indext> parent, ancestor, child, vertex, dom;
 
-  if(post_dom)
-    entry_node=cfg.get_last_node(program);
-  else
-    entry_node=cfg.get_first_node(program);
-  typename cfgt::nodet &n=cfg[cfg.entry_map[entry_node]];
-  n.dominators.insert(entry_node);
+    std::vector<node_indext> label, semi, size;
 
-  for(typename cfgt::edgest::const_iterator
-      s_it=(post_dom?n.in:n.out).begin();
-      s_it!=(post_dom?n.in:n.out).end();
-      ++s_it)
-    worklist.push_back(cfg[s_it->first].PC);
+    std::vector<std::set<node_indext>> bucket;
 
-  while(!worklist.empty())
-  {
-    // get node from worklist
-    T current=worklist.front();
-    worklist.pop_front();
+    T get_entry_node(P &program)
+    {
+      if(post_dom)
+      {
+        return cfg_dominators.cfg.get_last_node(program);
+      }
+      else
+      {
+        return cfg_dominators.cfg.get_first_node(program);
+      }
+    };
 
-    bool changed=false;
-    typename cfgt::nodet &node=cfg[cfg.entry_map[current]];
-    if(node.dominators.empty())
+    void dfs(node_indext root)
     {
-      for(const auto &edge : (post_dom ? node.out : node.in))
-        if(!cfg[edge.first].dominators.empty())
+      struct dfs_statet
+      {
+        node_indext parent;
+        node_indext current;
+      };
+      std::stack<dfs_statet> work;
+      work.push({0, root});
+      while(!work.empty())
+      {
+        auto state = work.top();
+        work.pop();
+        node_indext v = state.current;
+        if(semi[v] == 0)
         {
-          node.dominators=cfg[edge.first].dominators;
-          node.dominators.insert(current);
-          changed=true;
+          parent[v] = state.parent;
+          semi[v] = ++dfs_counter;
+          vertex[dfs_counter] = label[v] = v;
+          ancestor[v] = child[v] = 0;
+          size[v] = 1;
+          for_each_successor(v, [&](node_indext w) { work.push({v, w}); });
         }
+      }
     }
 
-    // compute intersection of predecessors
-    for(const auto &edge : (post_dom ? node.out : node.in))
+    void compress(node_indext v)
```

Comment:
I had feared as much: pretty much all of my previous comments are gone as they got killed by reformatting. Those comments were: where is the documentation?